lib/puppet/provider/user/directoryservice.rb

```diff
@@ -571,7 +571,32 @@ def set_shadow_hash_data(users_plist, binary_plist)
     else
       users_plist['ShadowHashData'] = [binary_plist]
     end
-    write_users_plist_to_disk(users_plist)
+    if Puppet::Util::Package.versioncmp(self.class.get_os_version, '10.15') < 0
+      write_users_plist_to_disk(users_plist)
+    else
+      write_and_import_shadow_hash_data(users_plist['ShadowHashData'].first)
+    end
+  end
+
+  # This method writes the ShadowHashData plist in a temporary file,
+  # then imports it using dsimport. macOS versions 10.15 and newer do
+  # not support directly managing binary plists, so we have to use an
+  # intermediary.
+  # dsimport is an archaic utilitary with hard-to-find documentation
+  #
+  # See http://web.archive.org/web/20090106120111/http://support.apple.com/kb/TA21305?viewlocale=en_US
+  # for information regarding the dsimport syntax
+  def write_and_import_shadow_hash_data(data_plist)
+    Tempfile.open("dsimport_#{@resource.name}", :encoding => Encoding::ASCII) do |dsimport_file|
+      dsimport_file.write <<-DSIMPORT
+0x0A 0x5C 0x3A 0x2C dsRecTypeStandard:Users 2 dsAttrTypeStandard:RecordName base64:dsAttrTypeNative:ShadowHashData
+#{@resource.name}:#{Base64.strict_encode64(data_plist)}
+      DSIMPORT
+      dsimport_file.flush
```

Comment:
if the block is closed here and dcsl command is executed outside I suppose flush is not needed anymore.

Reply:
true, I thought `Tempfile.open` deletes the file when the block closes, but looks like it only closes it and leaves the removal to the ruby garbage collector.

I'm not sure whether I should use `Tempfile.open`, put the dscl calls outside the block and then manually delete the file, or change to `Tempfile.create` and keep the current implementation...

Reply:
I want to have the block handle the file's lifetime as much as possible